# Damaris Campos PHW3: Chapter 4 & 5 + Propositions

## **<mark style=background-color:#f0c9db>Top 5 Queries:</mark>**

### **<font color=#b04c7a>TOP 1 - Chapter 4: Exercise 5</font>**

- Write a query that returns for each customer all orders placed on the customer's last day of activity
- Tables involved: TSQLV4 database, Orders table

Desired output:

| custid | orderid | orderdate | empid |
| --- | --- | --- | --- |
| 1 | 11011 | 2016-04-09 | 3 |
| 2 | 10926 | 2016-03-04 | 4 |
| 3 | 10856 | 2016-01-28 | 3 |
| 4 | 11016 | 2016-04-10 | 9 |
| 5 | 10924 | 2016-03-04 | 3 |
| ... | ... | ... | ... |
| 87 | 11025 | 2016-04-15 | 6 |
| 88 | 10935 | 2016-03-09 | 4 |
| 89 | 11066 | 2016-05-01 | 7 |
| 90 | 11005 | 2016-04-07 | 2 |
| 91 | 11044 | 2016-04-23 | 4 |

(90 rows affected)

In [5]:
--USE TSQLV4
--SELECT custid, orderid, orderdate, empid
--FROM Sales.Orders AS O1
--WHERE orderdate = (SELECT MAX(O2.orderdate) 
--                   FROM Sales.Orders AS O2
--                   WHERE O2.custid = O1.custid)
--ORDER BY custid;

USE Northwinds2022TSQLV7
SELECT CustomerId, OrderId, OrderDate, EmployeeId
FROM Sales.[Order] AS O1
WHERE OrderDate = (SELECT MAX(O2.OrderDate) 
                   FROM Sales.[Order] AS O2 
                   WHERE O2.CustomerId = O1.CustomerId)
ORDER BY CustomerId;

(90 rows affected)

Total execution time: 00:00:00.014

CustomerId OrderId OrderDate EmployeeId 1 11011 2016-04-09 3 2 10926 2016-03-04 4 3 10856 2016-01-28 3 4 11016 2016-04-10 9 5 10924 2016-03-04 3 6 11058 2016-04-29 9 7 10826 2016-01-12 6 8 10970 2016-03-24 9 9 11076 2016-05-06 4 10 11048 2016-04-24 7 11 11023 2016-04-14 1 12 11054 2016-04-28 8 13 10259 2014-07-18 4 14 11041 2016-04-22 3 15 11042 2016-04-22 2 16 10848 2016-01-23 7 17 11067 2016-05-04 1 18 10890 2016-02-16 7 19 11056 2016-04-28 8 20 11072 2016-05-05 4 21 10725 2015-10-31 4 23 10789 2015-12-22 1 24 11050 2016-04-27 8 25 11012 2016-04-09 1 26 10971 2016-03-24 2 27 11060 2016-04-30 2 28 10963 2016-03-19 9 29 10928 2016-03-05 1 30 11037 2016-04-21 7 31 11049 2016-04-24 3 32 11061 2016-04-30 4 33 10785 2015-12-18 1 34 11052 2016-04-27 3 35 11055 2016-04-28 7 36 10660 2015-09-08 8 37 11063 2016-04-30 3 38 10933 2016-03-06 6 39 11028 2016-04-16 2 40 10972 2016-03-24 4 40 10973 2016-03-24 6 41 11051 2016-04-27 7 42 10810 2016-01-01 2 43 10545 2015-05-22 8 44 11070 2016-05-05 2 45 10884 2016-02-12 4 46 11071 2016-05-05 1 47 11039 2016-04-21 1 48 11018 2016-04-13 4 49 10950 2016-03-16 1 50 11004 2016-04-07 3 51 10724 2015-10-30 8 52 10945 2016-03-12 4 53 11057 2016-04-29 3 54 10986 2016-03-30 8 55 11034 2016-04-20 8 56 11020 2016-04-14 2 58 11073 2016-05-05 2 59 11053 2016-04-27 2 60 11007 2016-04-08 8 61 10989 2016-03-31 2 62 11068 2016-05-04 8 63 11021 2016-04-14 3 64 11019 2016-04-13 6 65 11077 2016-05-06 1 66 11062 2016-04-30 4 67 11059 2016-04-29 2 68 11075 2016-05-06 8 69 11013 2016-04-09 2 70 11015 2016-04-10 2 71 11064 2016-05-01 1 72 10869 2016-02-04 5 73 11074 2016-05-06 7 74 11043 2016-04-22 5 75 10974 2016-03-25 3 76 11038 2016-04-21 1 77 10992 2016-04-01 1 78 11003 2016-04-06 3 79 10967 2016-03-23 2 80 11069 2016-05-04 1 81 10839 2016-01-19 3 82 10822 2016-01-08 6 83 10994 2016-04-02 2 84 10850 2016-01-23 1 85 10739 2015-11-12 3 86 11046 2016-04-23 8 87 11025 2016-04-15 6 88 10935 2016-03-09 4 89 11066 2016-05-01 7 90 11005 2016-04-07 2 91 11044 2016-04-23 4

**<font color=#b04c7a>Explanation:</font>**  
The proposition asks us to return information about the orders places on the last day of activity for each customer from the Sales.Order table. The columns selected are customerid, orderid, orderdate, and employeeid. These columns are obtained from our main table sales.order, which is aliased in the main query as 'O1'. The predicate we use is in our where clause. This filters the results to only include rows where the orderdate in the main table O1 is equal to the maximum orderdate for the same customer in the subquery O2. In the subquery, the MAX(O2.OrderDate) calculates the maximun order for each customer. The O2.CustomerId = O1.CustomerId condition makes sure the subquery is looking for maximum order date for the same customer in the main query. Finally, we order by customerid to get the ascending result.

### **<font color=#5e4cb0>TOP 2 - Chapter 5: Exercise 3-1</font>**

- Write a query that calculates a row number for each order based on orderdate, orderid ordering
- Tables involved: Sales.Orders

Desired output:

| orderid | orderdate | custid | empid | rownum |
| --- | --- | --- | --- | --- |
| 10248 | 2014-07-04 | 85 | 5 | 1 |
| 10249 | 2014-07-05 | 79 | 6 | 2 |
| 10250 | 2014-07-08 | 34 | 4 | 3 |
| 10251 | 2014-07-08 | 84 | 3 | 4 |
| 10252 | 2014-07-09 | 76 | 4 | 5 |
| 10253 | 2014-07-10 | 34 | 3 | 6 |
| 10254 | 2014-07-11 | 14 | 5 | 7 |
| 10255 | 2014-07-12 | 68 | 9 | 8 |
| 10256 | 2014-07-15 | 88 | 3 | 9 |
| 10257 | 2014-07-16 | 35 | 4 | 10 |
| ... | ... | ... | ... | ... |

In [26]:
--USE TSQLV4
--SELECT orderid, orderdate, custid, empid, ROW_NUMBER() OVER(ORDER BY orderdate, orderid) AS rownum 
--FROM Sales.Orders

USE Northwinds2022TSQLV7
SELECT OrderId, OrderDate, CustomerId, EmployeeId, ROW_NUMBER() OVER(ORDER BY OrderDate, OrderId) AS RowNum
FROM Sales.[Order]

(830 rows affected)

Total execution time: 00:00:00.030

OrderId OrderDate CustomerId EmployeeId RowNum 10248 2014-07-04 85 5 1 10249 2014-07-05 79 6 2 10250 2014-07-08 34 4 3 10251 2014-07-08 84 3 4 10252 2014-07-09 76 4 5 10253 2014-07-10 34 3 6 10254 2014-07-11 14 5 7 10255 2014-07-12 68 9 8 10256 2014-07-15 88 3 9 10257 2014-07-16 35 4 10 10258 2014-07-17 20 1 11 10259 2014-07-18 13 4 12 10260 2014-07-19 56 4 13 10261 2014-07-19 61 4 14 10262 2014-07-22 65 8 15 10263 2014-07-23 20 9 16 10264 2014-07-24 24 6 17 10265 2014-07-25 7 2 18 10266 2014-07-26 87 3 19 10267 2014-07-29 25 4 20 10268 2014-07-30 33 8 21 10269 2014-07-31 89 5 22 10270 2014-08-01 87 1 23 10271 2014-08-01 75 6 24 10272 2014-08-02 65 6 25 10273 2014-08-05 63 3 26 10274 2014-08-06 85 6 27 10275 2014-08-07 49 1 28 10276 2014-08-08 80 8 29 10277 2014-08-09 52 2 30 10278 2014-08-12 5 8 31 10279 2014-08-13 44 8 32 10280 2014-08-14 5 2 33 10281 2014-08-14 69 4 34 10282 2014-08-15 69 4 35 10283 2014-08-16 46 3 36 10284 2014-08-19 44 4 37 10285 2014-08-20 63 1 38 10286 2014-08-21 63 8 39 10287 2014-08-22 67 8 40 10288 2014-08-23 66 4 41 10289 2014-08-26 11 7 42 10290 2014-08-27 15 8 43 10291 2014-08-27 61 6 44 10292 2014-08-28 81 1 45 10293 2014-08-29 80 1 46 10294 2014-08-30 65 4 47 10295 2014-09-02 85 2 48 10296 2014-09-03 46 6 49 10297 2014-09-04 7 5 50 10298 2014-09-05 37 6 51 10299 2014-09-06 67 4 52 10300 2014-09-09 49 2 53 10301 2014-09-09 86 8 54 10302 2014-09-10 76 4 55 10303 2014-09-11 30 7 56 10304 2014-09-12 80 1 57 10305 2014-09-13 55 8 58 10306 2014-09-16 69 1 59 10307 2014-09-17 48 2 60 10308 2014-09-18 2 7 61 10309 2014-09-19 37 3 62 10310 2014-09-20 77 8 63 10311 2014-09-20 18 1 64 10312 2014-09-23 86 2 65 10313 2014-09-24 63 2 66 10314 2014-09-25 65 1 67 10315 2014-09-26 38 4 68 10316 2014-09-27 65 1 69 10317 2014-09-30 48 6 70 10318 2014-10-01 38 8 71 10319 2014-10-02 80 7 72 10320 2014-10-03 87 5 73 10321 2014-10-03 38 3 74 10322 2014-10-04 58 7 75 10323 2014-10-07 39 4 76 10324 2014-10-08 71 9 77 10325 2014-10-09 39 1 78 10326 2014-10-10 8 4 79 10327 2014-10-11 24 2 80 10328 2014-10-14 28 4 81 10329 2014-10-15 75 4 82 10330 2014-10-16 46 3 83 10331 2014-10-16 9 9 84 10332 2014-10-17 51 3 85 10333 2014-10-18 87 5 86 10334 2014-10-21 84 8 87 10335 2014-10-22 37 7 88 10336 2014-10-23 60 7 89 10337 2014-10-24 25 4 90 10338 2014-10-25 55 4 91 10339 2014-10-28 51 2 92 10340 2014-10-29 9 1 93 10341 2014-10-29 73 7 94 10342 2014-10-30 25 4 95 10343 2014-10-31 44 4 96 10344 2014-11-01 89 4 97 10345 2014-11-04 63 2 98 10346 2014-11-05 65 3 99 10347 2014-11-06 21 4 100 10348 2014-11-07 86 4 101 10349 2014-11-08 75 7 102 10350 2014-11-11 41 6 103 10351 2014-11-11 20 1 104 10352 2014-11-12 28 3 105 10353 2014-11-13 59 7 106 10354 2014-11-14 58 8 107 10355 2014-11-15 4 6 108 10356 2014-11-18 86 6 109 10357 2014-11-19 46 1 110 10358 2014-11-20 41 5 111 10359 2014-11-21 72 5 112 10360 2014-11-22 7 4 113 10361 2014-11-22 63 1 114 10362 2014-11-25 9 3 115 10363 2014-11-26 17 4 116 10364 2014-11-26 19 1 117 10365 2014-11-27 3 3 118 10366 2014-11-28 29 8 119 10367 2014-11-28 83 7 120 10368 2014-11-29 20 2 121 10369 2014-12-02 75 8 122 10370 2014-12-03 14 6 123 10371 2014-12-03 41 1 124 10372 2014-12-04 62 5 125 10373 2014-12-05 37 4 126 10374 2014-12-05 91 1 127 10375 2014-12-06 36 3 128 10376 2014-12-09 51 1 129 10377 2014-12-09 72 1 130 10378 2014-12-10 24 5 131 10379 2014-12-11 61 2 132 10380 2014-12-12 37 8 133 10381 2014-12-12 46 3 134 10382 2014-12-13 20 4 135 10383 2014-12-16 4 8 136 10384 2014-12-16 5 3 137 10385 2014-12-17 75 1 138 10386 2014-12-18 21 9 139 10387 2014-12-18 70 1 140 10388 2014-12-19 72 2 141 10389 2014-12-20 10 4 142 10390 2014-12-23 20 6 143 10391 2014-12-23 17 3 144 10392 2014-12-24 59 2 145 10393 2014-12-25 71 1 146 10394 2014-12-25 36 1 147 10395 2014-12-26 35 6 148 10396 2014-12-27 25 1 149 10397 2014-12-27 60 5 150 10398 2014-12-30 71 2 151 10399 2014-12-31 83 8 152 10400 2015-01-01 19 1 153 10401 2015-01-01 65 1 154 10402 2015-01-02 20 8 155 10403 2015-01-03 20 4 156 10404 2015-01-03 49

<font color=#5e4cb0>**Explanation:** </font><br>
The query uses the sales.order table to caluclate a row number for each other in the result set. The columns used to to be included in the result set was orderid, orderdate, customerid, and employeeid. We use the ROW_NUMBER() window function to assign unique row number in the results based on order by clause. In this case it provides orderdate and orderid in ascending manner. Here we are not using the where clause so no direct predicate, however, the row_number function uses a logical order to provide sequential numbering of rows.

## Propositions

### <font color=#4c79b0>**TOP 3 - Ch4 Q1:**</font>
- Return orders placed by employees whose last names start with the letter 'C'
- Tables involved: Sales.Orders

(104 rows affected)


In [7]:
-- Scalar subquery expected to return one value
-- USE TSQLV4
-- SELECT orderid
-- FROM Sales.Orders
-- WHERE empid = 
--  (SELECT E.empid
--   FROM HR.Employees AS E
--   WHERE E.lastname LIKE N'C%');
-- GO

USE Northwinds2022TSQLV7
SELECT OrderId
FROM Sales.[Order]
WHERE EmployeeId = 
    (SELECT E.EmployeeID
     FROM HumanResources.Employee AS E
     WHERE E.EmployeeLastName LIKE N'C%');
GO

(104 rows affected)

Total execution time: 00:00:00.018

OrderId 10262 10268 10276 10278 10279 10286 10287 10290 10301 10305 10310 10318 10334 10354 10366 10369 10380 10383 10399 10402 10408 10412 10416 10421 10435 10437 10443 10450 10452 10455 10456 10460 10467 10472 10476 10481 10488 10491 10498 10521 10533 10534 10543 10545 10560 10565 10571 10589 10596 10602 10603 10610 10614 10623 10627 10631 10632 10635 10651 10660 10679 10694 10696 10706 10719 10720 10722 10724 10729 10756 10770 10786 10795 10811 10824 10844 10845 10852 10857 10862 10883 10887 10932 10940 10955 10957 10961 10962 10977 10979 10986 10987 10997 10998 11007 11034 11036 11046 11050 11054 11056 11065 11068 11075

<font color=#4c79b0>**Explanation:**</font> <br>
The proposition asks us to identify and return orders placed by employees whose last names start with the letter 'C. The main table we use is sales.order and the column selected for that is orderid. The subquery involves the humanresources.employee table with the columns employeeid and employeelastname. The predicate is in our where clause where it is both in the main and subquery. In the subquery, it is returning employeeid for employees whose last names start with 'C' by using LIKE N'C' predicate so only employees with last names beginning with 'C' are selected. Main query where clause condition filters results to include only oders placed by employees meeting the criteria specified in the subquery.

### <font color=#4ca3b0>**TOP 4 - Ch4 Q2:**</font>
- Return information about each order from the Sales.[Order] table along with subsequent order's ID ('NextOrderId')
- Tables involved: Sales.Orders

In [9]:
-- USE TSQLV4
-- SELECT orderid, orderdate, empid, custid,
--  (SELECT MIN(O2.orderid)
--   FROM Sales.Orders AS O2
--   WHERE O2.orderid > O1.orderid) AS nextorderid
-- FROM Sales.Orders AS O1;

USE Northwinds2022TSQLV7
SELECT OrderId, OrderDate, EmployeeId, CustomerId,
    (SELECT MIN(O2.OrderId)
     FROM Sales.[Order] AS O2
     WHERE O2.OrderId > O1.OrderId) AS NextOrderId
FROM Sales.[Order] AS O1;


(830 rows affected)

Total execution time: 00:00:00.011

OrderId OrderDate EmployeeId CustomerId NextOrderId 10248 2014-07-04 5 85 10249 10249 2014-07-05 6 79 10250 10250 2014-07-08 4 34 10251 10251 2014-07-08 3 84 10252 10252 2014-07-09 4 76 10253 10253 2014-07-10 3 34 10254 10254 2014-07-11 5 14 10255 10255 2014-07-12 9 68 10256 10256 2014-07-15 3 88 10257 10257 2014-07-16 4 35 10258 10258 2014-07-17 1 20 10259 10259 2014-07-18 4 13 10260 10260 2014-07-19 4 56 10261 10261 2014-07-19 4 61 10262 10262 2014-07-22 8 65 10263 10263 2014-07-23 9 20 10264 10264 2014-07-24 6 24 10265 10265 2014-07-25 2 7 10266 10266 2014-07-26 3 87 10267 10267 2014-07-29 4 25 10268 10268 2014-07-30 8 33 10269 10269 2014-07-31 5 89 10270 10270 2014-08-01 1 87 10271 10271 2014-08-01 6 75 10272 10272 2014-08-02 6 65 10273 10273 2014-08-05 3 63 10274 10274 2014-08-06 6 85 10275 10275 2014-08-07 1 49 10276 10276 2014-08-08 8 80 10277 10277 2014-08-09 2 52 10278 10278 2014-08-12 8 5 10279 10279 2014-08-13 8 44 10280 10280 2014-08-14 2 5 10281 10281 2014-08-14 4 69 10282 10282 2014-08-15 4 69 10283 10283 2014-08-16 3 46 10284 10284 2014-08-19 4 44 10285 10285 2014-08-20 1 63 10286 10286 2014-08-21 8 63 10287 10287 2014-08-22 8 67 10288 10288 2014-08-23 4 66 10289 10289 2014-08-26 7 11 10290 10290 2014-08-27 8 15 10291 10291 2014-08-27 6 61 10292 10292 2014-08-28 1 81 10293 10293 2014-08-29 1 80 10294 10294 2014-08-30 4 65 10295 10295 2014-09-02 2 85 10296 10296 2014-09-03 6 46 10297 10297 2014-09-04 5 7 10298 10298 2014-09-05 6 37 10299 10299 2014-09-06 4 67 10300 10300 2014-09-09 2 49 10301 10301 2014-09-09 8 86 10302 10302 2014-09-10 4 76 10303 10303 2014-09-11 7 30 10304 10304 2014-09-12 1 80 10305 10305 2014-09-13 8 55 10306 10306 2014-09-16 1 69 10307 10307 2014-09-17 2 48 10308 10308 2014-09-18 7 2 10309 10309 2014-09-19 3 37 10310 10310 2014-09-20 8 77 10311 10311 2014-09-20 1 18 10312 10312 2014-09-23 2 86 10313 10313 2014-09-24 2 63 10314 10314 2014-09-25 1 65 10315 10315 2014-09-26 4 38 10316 10316 2014-09-27 1 65 10317 10317 2014-09-30 6 48 10318 10318 2014-10-01 8 38 10319 10319 2014-10-02 7 80 10320 10320 2014-10-03 5 87 10321 10321 2014-10-03 3 38 10322 10322 2014-10-04 7 58 10323 10323 2014-10-07 4 39 10324 10324 2014-10-08 9 71 10325 10325 2014-10-09 1 39 10326 10326 2014-10-10 4 8 10327 10327 2014-10-11 2 24 10328 10328 2014-10-14 4 28 10329 10329 2014-10-15 4 75 10330 10330 2014-10-16 3 46 10331 10331 2014-10-16 9 9 10332 10332 2014-10-17 3 51 10333 10333 2014-10-18 5 87 10334 10334 2014-10-21 8 84 10335 10335 2014-10-22 7 37 10336 10336 2014-10-23 7 60 10337 10337 2014-10-24 4 25 10338 10338 2014-10-25 4 55 10339 10339 2014-10-28 2 51 10340 10340 2014-10-29 1 9 10341 10341 2014-10-29 7 73 10342 10342 2014-10-30 4 25 10343 10343 2014-10-31 4 44 10344 10344 2014-11-01 4 89 10345 10345 2014-11-04 2 63 10346 10346 2014-11-05 3 65 10347 10347 2014-11-06 4 21 10348 10348 2014-11-07 4 86 10349 10349 2014-11-08 7 75 10350 10350 2014-11-11 6 41 10351 10351 2014-11-11 1 20 10352 10352 2014-11-12 3 28 10353 10353 2014-11-13 7 59 10354 10354 2014-11-14 8 58 10355 10355 2014-11-15 6 4 10356 10356 2014-11-18 6 86 10357 10357 2014-11-19 1 46 10358 10358 2014-11-20 5 41 10359 10359 2014-11-21 5 72 10360 10360 2014-11-22 4 7 10361 10361 2014-11-22 1 63 10362 10362 2014-11-25 3 9 10363 10363 2014-11-26 4 17 10364 10364 2014-11-26 1 19 10365 10365 2014-11-27 3 3 10366 10366 2014-11-28 8 29 10367 10367 2014-11-28 7 83 10368 10368 2014-11-29 2 20 10369 10369 2014-12-02 8 75 10370 10370 2014-12-03 6 14 10371 10371 2014-12-03 1 41 10372 10372 2014-12-04 5 62 10373 10373 2014-12-05 4 37 10374 10374 2014-12-05 1 91 10375 10375 2014-12-06 3 36 10376 10376 2014-12-09 1 51 10377 10377 2014-12-09 1 72 10378 10378 2014-12-10 5 24 10379 10379 2014-12-11 2 61 10380 10380 2014-12-12 8 37 10381 10381 2014-12-12 3 46 10382 10382 2014-12-13 4 20 10383 10383 2014-12-16 8 4 10384 10384 2014-12-16 3 5 10385 10385 2014-12-17 1 75 10386 10386 2014-12-18 9 21 10387 10387 2014-12-18 1 70 10388 10388 2014-12-19 2 72 10389 10389 2014-12-

<font color=#4ca3b0>**Explanation:**</font><br>
The proposition for this query is to return information about each order in the Sales.[Order] table. It includes columns such as orderid, orderdate, employeeid, and customerid. There is also a column called nextorderid where it shows the minimum order id for the next order in the sequence which we use a subquery to achieve this. This provides the minimum order ID greater than the current order's ID. The predicate is within the subquery, specifically WHERE O2.OrderId > O1.OrderId. This condition filters orders from O2 where the order ID is greater than the order ID in the outer query (O1). The subquery retrieves the minimum order ID (MIN(O2.OrderId)) from the Sales.[Order] table (O2) where the order ID is greater than the current order's ID in the outer query (O1). This finds the next order ID for each order.


### <font color=#4cb074>**TOP 5 - Ch4 Q3:**</font>
- Return customers from Spain that have placed at least one order.
- Tables involved: Sales.Customers

In [8]:
-- Customers from Spain who placed orders
-- USE TSQLV4
-- SELECT custid, companyname
-- FROM Sales.Customers AS C
-- WHERE country = N'Spain'
--   AND EXISTS
--     (SELECT * FROM Sales.Orders AS O
--      WHERE O.custid = C.custid);

USE Northwinds2022TSQLV7
SELECT CustomerId, CustomerCompanyName
FROM Sales.Customer AS C
WHERE CustomerCountry = N'Spain'
    AND EXISTS 
        (SELECT * FROM Sales.[Order] AS O 
         WHERE O.CustomerId = C.CustomerId);



(4 rows affected)

Total execution time: 00:00:00.022

CustomerId CustomerCompanyName 8 Customer QUHWH 29 Customer MDLWA 30 Customer KSLQF 69 Customer SIUIH

<font color=#4cb074>**Explanation:**</font><br>
The query provides information about customers in Spain who have made purchases. The main table is Sales.Customer and the columns selected are CustomerId and CustomerCompanyName. The subquery uses the Sales.[Order] table and uses the * to check for all records there. The predicate is in the WHERE clause, specifically WHERE CustomerCountry = N'Spain'. This filters the customers based on their country. The EXISTS clause introduces a subquery with the predicate WHERE O.CustomerId = C.CustomerId. This ensures that only customers who have corresponding records in the Sales.[Order] table are included in the results. The subquery checks for the existence of any records in the Sales.[Order] table where the CustomerId matches the CustomerId of the outer query's Sales.Customer table. The main query selects the CustomerId and CustomerCompanyName for customers located in Spain (CustomerCountry = N'Spain') who have placed at least one order.

**Ch4 Q4:**</font>
- Return a list of missing order IDs within the specified range by creating a temporary table that contains even order IDs.
- Tables involved: dbo.Orders, dbo.Nums

(414 rows affected)

In [19]:
-- Missing order IDs
--USE TSQLV4;
--DROP TABLE IF EXISTS dbo.Orders;
--CREATE TABLE dbo.Orders(orderid INT NOT NULL CONSTRAINT PK_Orders PRIMARY KEY);

--INSERT INTO dbo.Orders(orderid)
--  SELECT orderid
--  FROM Sales.Orders
--  WHERE orderid % 2 = 0;

USE Northwinds2022TSQLV7;
DROP TABLE IF EXISTS dbo.Orders;
CREATE TABLE dbo.Orders(orderid INT NOT NULL CONSTRAINT PK_Orders PRIMARY KEY);

INSERT INTO dbo.Orders(orderid)
  SELECT OrderID
  FROM Sales.[Order]
  WHERE OrderId % 2 = 0;

SELECT n
FROM dbo.Nums
WHERE n BETWEEN (SELECT MIN(O.OrderId) FROM dbo.Orders AS O)
 AND (SELECT MAX(O.OrderId) FROM dbo.Orders AS O)
 AND n NOT IN (SELECT O.OrderId FROM dbo.Orders AS O);

(415 rows affected)

(414 rows affected)

Total execution time: 00:00:00.074

n 10249 10251 10253 10255 10257 10259 10261 10263 10265 10267 10269 10271 10273 10275 10277 10279 10281 10283 10285 10287 10289 10291 10293 10295 10297 10299 10301 10303 10305 10307 10309 10311 10313 10315 10317 10319 10321 10323 10325 10327 10329 10331 10333 10335 10337 10339 10341 10343 10345 10347 10349 10351 10353 10355 10357 10359 10361 10363 10365 10367 10369 10371 10373 10375 10377 10379 10381 10383 10385 10387 10389 10391 10393 10395 10397 10399 10401 10403 10405 10407 10409 10411 10413 10415 10417 10419 10421 10423 10425 10427 10429 10431 10433 10435 10437 10439 10441 10443 10445 10447 10449 10451 10453 10455 10457 10459 10461 10463 10465 10467 10469 10471 10473 10475 10477 10479 10481 10483 10485 10487 10489 10491 10493 10495 10497 10499 10501 10503 10505 10507 10509 10511 10513 10515 10517 10519 10521 10523 10525 10527 10529 10531 10533 10535 10537 10539 10541 10543 10545 10547 10549 10551 10553 10555 10557 10559 10561 10563 10565 10567 10569 10571 10573 10575 10577 10579 10581 10583 10585 10587 10589 10591 10593 10595 10597 10599 10601 10603 10605 10607 10609 10611 10613 10615 10617 10619 10621 10623 10625 10627 10629 10631 10633 10635 10637 10639 10641 10643 10645 10647 10649 10651 10653 10655 10657 10659 10661 10663 10665 10667 10669 10671 10673 10675 10677 10679 10681 10683 10685 10687 10689 10691 10693 10695 10697 10699 10701 10703 10705 10707 10709 10711 10713 10715 10717 10719 10721 10723 10725 10727 10729 10731 10733 10735 10737 10739 10741 10743 10745 10747 10749 10751 10753 10755 10757 10759 10761 10763 10765 10767 10769 10771 10773 10775 10777 10779 10781 10783 10785 10787 10789 10791 10793 10795 10797 10799 10801 10803 10805 10807 10809 10811 10813 10815 10817 10819 10821 10823 10825 10827 10829 10831 10833 10835 10837 10839 10841 10843 10845 10847 10849 10851 10853 10855 10857 10859 10861 10863 10865 10867 10869 10871 10873 10875 10877 10879 10881 10883 10885 10887 10889 10891 10893 10895 10897 10899 10901 10903 10905 10907 10909 10911 10913 10915 10917 10919 10921 10923 10925 10927 10929 10931 10933 10935 10937 10939 10941 10943 10945 10947 10949 10951 10953 10955 10957 10959 10961 10963 10965 10967 10969 10971 10973 10975 10977 10979 10981 10983 10985 10987 10989 10991 10993 10995 10997 10999 11001 11003 11005 11007 11009 11011 11013 11015 11017 11019 11021 11023 11025 11027 11029 11031 11033 11035 11037 11039 11041 11043 11045 11047 11049 11051 11053 11055 11057 11059 11061 11063 11065 11067 11069 11071 11073 11075

In [4]:
-- CLeanup
DROP TABLE IF EXISTS dbo.Orders;

Commands completed successfully.

Total execution time: 00:00:00.005

**Explanation:** <br>
The proposition asks that we create a temporary table, populate it with even order IDs from the Sales.[Order] table, and then identifies missing order IDs within a specified range using the dbo.Nums table. The main table involved is the temporary table dbo.Orders that is created with the column orderid. The predicate is the where orderid % 2= 0 where it filters even order IDs from the Sales.[Order] table. Since the query that populated the dbo.orders table filtered even orders, this query returns all odd values between minimum and maximum order ID's. The subqueries are used in the range selection part of the main query. (SELECT MIN(O.orderid) FROM dbo.Orders AS O) finds the minimum order ID in the temporary dbo.Orders table, and (SELECT MAX(O.orderid) FROM dbo.Orders AS O) finds the maximum order ID. The main query selects values from the dbo.Nums table column n. It restricts the selection to values between the minimum and maximum order IDs in the temporary dbo.Orders table and excludes those that already exist in the dbo.Orders table. This identifies missing order IDs within the specified range. Finally, the cleanup section ensures that the temporary table is dropped at the end.

**Ch4 Q5:**

- Return customers who have not placed any orders along with the company name
- Tables involved: Sales.Customers

In [10]:
-- Using NOT EXISTS
-- USE TSQLV4
-- SELECT custid, companyname
-- FROM Sales.Customers AS C
-- WHERE NOT EXISTS
--   (SELECT * 
--    FROM Sales.Orders AS O
--    WHERE O.custid = C.custid);

USE Northwinds2022TSQLV7
SELECT CustomerId, CustomerCompanyName
FROM Sales.Customer AS C
WHERE NOT EXISTS 
    (SELECT *
     FROM Sales.[Order] AS O 
     WHERE O.CustomerId = C.CustomerId);


(2 rows affected)

Total execution time: 00:00:00.007

CustomerId CustomerCompanyName 22 Customer DTDMN 57 Customer WVAXS

In [11]:
--USE TSQLV4
-- Cleanup
-- DELETE FROM Sales.Orders WHERE custid IS NULL;
-- GO

USE Northwinds2022TSQLV7
DELETE FROM Sales.[Order] WHERE CustomerId IS NULL;
GO

(0 rows affected)

Total execution time: 00:00:00.005

**Explanation:** <br>The query identifies and returns customers from the Sales.Customer table who have not placed any orders. The main table involved is Sales.Customer, and the columns selected are CustomerId and CustomerCompanyName. The subquery uses the Sales.[Order] table and uses the * to check for the existence of any records. The main predicate is in the WHERE clause, specifically WHERE NOT EXISTS. This checks for the absence of records based on the condition in the subquery. The subquery contains a predicate in the form of WHERE O.CustomerId = C.CustomerId. This ensures that the subquery is looking for orders with the same customer as in the main query. The subquery checks for the existence of any records in the Sales.[Order] table where the CustomerId matches the CustomerId of the outer query's Sales.Customer table. The NOT EXISTS condition ensures that only customers without corresponding orders are included in the result set. 

**Ch5 Q6:**

- Extract order year, current year customer count, previous year customer count to calculate the growth in customer count
- Tables involved: Sales.Orders

In [12]:
-- USE TSQLV4;
-- WITH YearlyCount AS
-- (
--  SELECT YEAR(orderdate) AS orderyear,
--    COUNT(DISTINCT custid) AS numcusts
--  FROM Sales.Orders
--  GROUP BY YEAR(orderdate)
-- )
-- SELECT Cur.orderyear, 
--  Cur.numcusts AS curnumcusts, Prv.numcusts AS prvnumcusts,
--  Cur.numcusts - Prv.numcusts AS growth
-- FROM YearlyCount AS Cur
--  LEFT OUTER JOIN YearlyCount AS Prv
--    ON Cur.orderyear = Prv.orderyear + 1;


USE Northwinds2022TSQLV7;

WITH YearlyCount AS
(
  SELECT YEAR(OrderDate) AS OrderYear,
    COUNT(DISTINCT CustomerId) AS NumCusts
  FROM Sales.[Order]
  GROUP BY YEAR(OrderDate)
)
SELECT Cur.OrderYear, 
  Cur.NumCusts AS CurNumCusts, Prv.NumCusts AS PrvNumCusts,
  Cur.NumCusts - Prv.NumCusts AS Growth
FROM YearlyCount AS Cur
  LEFT OUTER JOIN YearlyCount AS Prv
    ON Cur.OrderYear = Prv.OrderYear + 1;

(3 rows affected)

Total execution time: 00:00:00.013

OrderYear CurNumCusts PrvNumCusts Growth 2014 67 NULL NULL 2015 86 67 19 2016 81 86 -5

**Explanation:**  
This query utilizes a Common Table Expression (CTE) named YearlyCount to compute the yearly count of distinct customers based on order dates in the Sales.\[Order\] table. The columns are orderyear, numcusts which indicates distinct count for each year, curnumcusts as the current year's customer count, prvnumcusts indicated as previous year's customer count, and growth that computes difference between current and previous years. The correlation between the 'Cur' and 'Prv' instances in the LEFT OUTER JOIN is based on the condition 'Cur.OrderYear = Prv.OrderYear + 1.' This ensures that the counts from the current year are matched with the counts from the previous year for the growth calculation.

**Ch5 Q7:**

- Extract order year and calculate the distinct customer count for the specified employee (@empid) for each year
- Tables involved: Sales.Orders

In [13]:
-- USE TSQLV4
-- DECLARE @empid AS INT = 3;

-- WITH C AS
-- (
--  SELECT YEAR(orderdate) AS orderyear, custid
--  FROM Sales.Orders
--  WHERE empid = @empid
-- )
-- SELECT orderyear, COUNT(DISTINCT custid) AS numcusts
-- FROM C
-- GROUP BY orderyear;
-- GO

USE Northwinds2022TSQLV7
DECLARE @empid AS INT = 3;

WITH C AS
(
    SELECT YEAR(OrderDate) AS OrderYear, CustomerId
    FROM Sales.[Order]
    WHERE EmployeeId = @empid
)
SELECT OrderYear, COUNT(DISTINCT CustomerId) AS NumCusts 
FROM C
GROUP BY OrderYear;
GO

(3 rows affected)

Total execution time: 00:00:00.012

OrderYear NumCusts 2014 16 2015 46 2016 30

**Explanation:**  
The proposition wants us to analyze the customer count for a specific employee ID (@empid) over different years. The main table involved is Sales.[Order]. The columns selected in the CTE (C) are YEAR(OrderDate) as OrderYear and CustomerId. The main query extracts OrderYear and calculates the distinct count of customers (NumCusts) from the CTE. The predicate is within the CTE, specifically WHERE EmployeeId = @empid. This condition filters orders for the specified employee ID. The CTE named C extracts the order year and customer ID for orders associated with the specified employee ID (@empid). The GROUP BY OrderYear groups the results by the order year.

**Ch5 Q8:**

- Extract order years and calculate the distinct customer count for each year from the Sales.\[Order\] table. Group the results by order year and include only the years where the distinct customer count exceeds 70
- Tables involved: Sales.Orders

In [14]:
-- USE TSQLV4
-- SELECT YEAR(orderdate) AS orderyear, COUNT(DISTINCT custid) AS numcusts
-- FROM Sales.Orders
-- GROUP BY YEAR(orderdate)
-- HAVING COUNT(DISTINCT custid) > 70;

USE Northwinds2022TSQLV7
SELECT YEAR(OrderDate) AS OrderYear, COUNT(DISTINCT CustomerId) AS NumCusts 
FROM Sales.[Order]
GROUP BY YEAR(OrderDate)
HAVING COUNT(DISTINCT CustomerId) > 70

(2 rows affected)

Total execution time: 00:00:00.006

OrderYear NumCusts 2015 86 2016 81

**Explanation:**  
The query aims to identify and summarize the order years with a significant number of distinct customers, focusing on years where the distinct customer count is greater than 70. The main table involved is Sales.[Order]. The columns selected are YEAR(OrderDate) AS OrderYear to extract the order year and COUNT(DISTINCT CustomerId) AS NumCusts to calculate the distinct customer count for each year. The primary predicate is within the HAVING clause: HAVING COUNT(DISTINCT CustomerId) > 70. This condition filters the grouped results to include only those order years where the distinct customer count is greater than 70. The GROUP BY YEAR(OrderDate) groups the results by the order year.

**Ch5 Q9:**

- Retrieve the customer ID, order ID, and order date for each customer from the Sales.Customer table. Utilize the OUTER APPLY operator to associate each customer with the top three most recent orders
- Tables involved: Sales.Customers

In [15]:
-- USE TSQLV4;
-- SELECT C.custid, A.orderid, A.orderdate
-- FROM Sales.Customers AS C
--  OUTER APPLY
--  (SELECT TOP (3) orderid, empid, orderdate, requireddate
--  FROM Sales.Orders AS O
--  WHERE O.custid = C.custid
--  ORDER BY orderdate DESC, orderid DESC) AS A;

USE Northwinds2022TSQLV7
SELECT C.CustomerID, A.OrderId, A.OrderDate
FROM Sales.Customer AS C
 OUTER APPLY
 (SELECT TOP (3) OrderId, EmployeeId, OrderDate, RequiredDate
 FROM Sales.[Order] AS O
 WHERE O.CustomerId = C.CustomerId
 ORDER BY OrderDate DESC, OrderId DESC) AS A;

(265 rows affected)

Total execution time: 00:00:00.026

CustomerID OrderId OrderDate 1 11011 2016-04-09 1 10952 2016-03-16 1 10835 2016-01-15 2 10926 2016-03-04 2 10759 2015-11-28 2 10625 2015-08-08 3 10856 2016-01-28 3 10682 2015-09-25 3 10677 2015-09-22 4 11016 2016-04-10 4 10953 2016-03-16 4 10920 2016-03-03 5 10924 2016-03-04 5 10875 2016-02-06 5 10866 2016-02-03 6 11058 2016-04-29 6 10956 2016-03-17 6 10853 2016-01-27 7 10826 2016-01-12 7 10679 2015-09-23 7 10628 2015-08-12 8 10970 2016-03-24 8 10801 2015-12-29 8 10326 2014-10-10 9 11076 2016-05-06 9 10940 2016-03-11 9 10932 2016-03-06 10 11048 2016-04-24 10 11045 2016-04-23 10 11027 2016-04-16 11 11023 2016-04-14 11 10947 2016-03-13 11 10943 2016-03-11 12 11054 2016-04-28 12 10937 2016-03-10 12 10881 2016-02-11 13 10259 2014-07-18 14 11041 2016-04-22 14 11029 2016-04-16 14 10966 2016-03-20 15 11042 2016-04-22 15 10969 2016-03-23 15 10494 2015-04-02 16 10848 2016-01-23 16 10462 2015-03-03 16 10435 2015-02-04 17 11067 2016-05-04 17 11036 2016-04-20 17 10825 2016-01-09 18 10890 2016-02-16 18 10683 2015-09-26 18 10609 2015-07-24 19 11056 2016-04-28 19 11047 2016-04-24 19 11024 2016-04-15 20 11072 2016-05-05 20 11017 2016-04-13 20 11008 2016-04-08 21 10725 2015-10-31 21 10650 2015-08-29 21 10581 2015-06-26 22 NULL NULL 23 10789 2015-12-22 23 10763 2015-12-03 23 10634 2015-08-15 24 11050 2016-04-27 24 11001 2016-04-06 24 10993 2016-04-01 25 11012 2016-04-09 25 10929 2016-03-05 25 10859 2016-01-29 26 10971 2016-03-24 26 10860 2016-01-29 26 10671 2015-09-17 27 11060 2016-04-30 27 11026 2016-04-15 27 10807 2015-12-31 28 10963 2016-03-19 28 10664 2015-09-10 28 10604 2015-07-18 29 10928 2016-03-05 29 10887 2016-02-13 29 10568 2015-06-13 30 11037 2016-04-21 30 11009 2016-04-08 30 10948 2016-03-13 31 11049 2016-04-24 31 10959 2016-03-18 31 10790 2015-12-22 32 11061 2016-04-30 32 11040 2016-04-22 32 11006 2016-04-07 33 10785 2015-12-18 33 10268 2014-07-30 34 11052 2016-04-27 34 11022 2016-04-14 34 10981 2016-03-27 35 11055 2016-04-28 35 10976 2016-03-25 35 10960 2016-03-19 36 10660 2015-09-08 36 10600 2015-07-16 36 10415 2015-01-15 37 11063 2016-04-30 37 10985 2016-03-30 37 10912 2016-02-26 38 10933 2016-03-06 38 10829 2016-01-13 38 10798 2015-12-26 39 11028 2016-04-16 39 10893 2016-02-18 39 10849 2016-01-23 40 10973 2016-03-24 40 10972 2016-03-24 40 10927 2016-03-05 41 11051 2016-04-27 41 10923 2016-03-03 41 10832 2016-01-14 42 10810 2016-01-01 42 10620 2015-08-05 42 10495 2015-04-03 43 10545 2015-05-22 43 10482 2015-03-21 44 11070 2016-05-05 44 10934 2016-03-09 44 10891 2016-02-17 45 10884 2016-02-12 45 10735 2015-11-10 45 10719 2015-10-27 46 11071 2016-05-05 46 11065 2016-05-01 46 10997 2016-04-03 47 11039 2016-04-21 47 11014 2016-04-10 47 10954 2016-03-17 48 11018 2016-04-13 48 10883 2016-02-12 48 10867 2016-02-03 49 10950 2016-03-16 49 10939 2016-03-10 49 10818 2016-01-07 50 11004 2016-04-07 50 10978 2016-03-26 50 10896 2016-02-19 51 10724 2015-10-30 51 10619 2015-08-04 51 10618 2015-08-01 52 10945 2016-03-12 52 10779 2015-12-16 52 10699 2015-10-09 53 11057 2016-04-29 53 10752 2015-11-24 53 10517 2015-04-24 54 10986 2016-03-30 54 10958 2016-03-18 54 10898 2016-02-20 55 11034 2016-04-20 55 10965 2016-03-20 55 10855 2016-01-27 56 11020 2016-04-14 56 10999 2016-04-03 56 10833 2016-01-15 57 NULL NULL 58 11073 2016-05-05 58 10995 2016-04-02 58 10502 2015-04-10 59 11053 2016-04-27 59 10844 2016-01-21 59 10747 2015-11-19 60 11007 2016-04-08 60 10477 2015-03-17 60 10433 2015-02-03 61 10989 2016-03-31 61 10794 2015-12-24 61 10720 2015-10-28 62 11068 2016-05-04 62 10961 2016-03-19 62 10914 2016-02-27 63 11021 2016-04-14 63 10996 2016-04-02 63 10991 2016-04-01 64 11019 2016-04-13 64 10916 2016-02-27 64 10828 2016-01-13 65 11077 2016-05-06 65 11000 2016-04-06 65 10988 2016-03-31 66 11062 2016-04-30 66 11010 2016-04-09 66 10942 2016-03-11 67 11059 2016-04-29 67 10877 2016-02-09 67 10851 2016-01-26 68 11075 2016-05-06 68 11033 2016-04-17 68 10951 2016-03-16 69 11013 2016-04-09 69 10917 2016-03-02 69 10306 2014-09-16 70 11015 2016-04-10 70 10909 201

**Explanation:**  
The proposition asks us to retrieve information for each customer from the Sales.Customer table along with details of their top three most recent orders. The main table involved is Sales.Customer, and the columns selected for display are CustomerID. The subquery within the OUTER APPLY clause involves the Sales.[Order] table and selects columns such as OrderId, EmployeeId, OrderDate, and RequiredDate. The OUTER APPLY operator is used to associate each row from the Sales.Customer table (C) with the result of the correlated subquery (A). The predicate is within the subquery WHERE O.CustomerId = C.CustomerId. This condition ensures that the orders selected in the subquery are associated with the corresponding customer in the outer query.

**Ch5 Q10:**

- Create a view named 'Sales.USACusts' in the 'TSQLV4' database to selectively include customers from the USA based on specified columns from the 'Sales.Customers' table.
- Return customer IDs and company names for USA-based customers through the view.

In [25]:
--USE TSQLV4
--DROP VIEW IF EXISTS Sales.USACusts;
--GO
--CREATE VIEW Sales.USACusts
--AS

--SELECT
--  custid, companyname, contactname, contacttitle, address,
--  city, region, postalcode, country, phone, fax
--FROM Sales.Customers
--WHERE country = N'USA';
--GO

--SELECT custid, companyname
--FROM Sales.USACusts;
--GO
-- Notice that you can insert a row through the view
--INSERT INTO Sales.USACusts(
--  companyname, contactname, contacttitle, address,
--  city, region, postalcode, country, phone, fax)
-- VALUES(
--  N'Customer ABCDE', N'Contact ABCDE', N'Title ABCDE', N'Address ABCDE',
--  N'London', NULL, N'12345', N'UK', N'012-3456789', N'012-3456789');

-- But when you query the view, you won't see it
--SELECT custid, companyname, country
--FROM Sales.USACusts
--WHERE companyname = N'Customer ABCDE';

-- You can see it in the table, though
--SELECT custid, companyname, country
--FROM Sales.Customers
--WHERE companyname = N'Customer ABCDE';
--GO


USE Northwinds2022TSQLV7
DROP VIEW IF EXISTS Sales.USACusts;
GO
CREATE VIEW Sales.USACusts
AS

SELECT
  CustomerId, CustomerCompanyName, CustomerContactName, CustomerContactTitle, CustomerAddress,
  CustomerCity, CustomerRegion, CustomerPostalCode, CustomerCountry, CustomerPhoneNumber, CustomerFaxNumber
FROM Sales.Customer
WHERE CustomerCountry = N'USA';
GO

SELECT CustomerId, CustomerCompanyName
FROM Sales.USACusts;
GO
-- Notice that you can insert a row through the view
INSERT INTO Sales.USACusts(
  CustomerCompanyName, CustomerContactName, CustomerContactTitle, CustomerAddress,
  CustomerCity, CustomerRegion, CustomerPostalCode, CustomerCountry, CustomerPhoneNumber, CustomerFaxNumber)
 VALUES(
  N'Customer ABCDE', N'Contact ABCDE', N'Title ABCDE', N'Address ABCDE',
  N'London', NULL, N'12345', N'UK', N'012-3456789', N'012-3456789');

-- But when you query the view, you won't see it
SELECT CustomerId, CustomerCompanyName, CustomerCountry
FROM Sales.USACusts
WHERE CustomerCompanyName = N'Customer ABCDE';

-- You can see it in the table, though
SELECT CustomerId, CustomerCompanyName, CustomerCountry
FROM Sales.Customer
WHERE CustomerCompanyName = N'Customer ABCDE';
GO

Commands completed successfully.

Commands completed successfully.

(13 rows affected)

(1 row affected)

(0 rows affected)

(1 row affected)

Total execution time: 00:00:00.067

CustomerId CustomerCompanyName 32 Customer YSIQX 36 Customer LVJSO 43 Customer UISOJ 45 Customer QXPPT 48 Customer DVFMB 55 Customer KZQZT 65 Customer NYUHS 71 Customer LCOUJ 75 Customer XOJYP 77 Customer LCYBZ 78 Customer NLTYP 82 Customer EYHKM 89 Customer YBQTI

CustomerId CustomerCompanyName CustomerCountry

CustomerId CustomerCompanyName CustomerCountry 92 Customer ABCDE UK

**Explanation:**  
The proposition asks us to creates a view to selectively display customers from the USA and demonstrates the insertion of a new row directly into the view, but when querying the view, the inserted row remains hidden due to the 'USA' filtering condition. The table involved is Sales.Customers. Columns selected for the view include custid, companyname, contactname, contacttitle, address, city, region, postalcode, country, phone, and fax. The predicate is within the view's definition: WHERE country = N'USA'. This condition filters customers based on their country, selectively including only those from the USA. The CREATE VIEW statement establishes a view named 'Sales.USACusts' that encapsulates the logic of selecting specific columns from the 'Sales.Customers' table where the country is 'USA'. The script queries the view 'Sales.USACusts' to retrieve custid and companyname for USA-based customers. Demonstrates the insertion of a new row directly into the view 'Sales.USACusts'. However, the inserted row is not visible when querying the view due to the filtering condition. The query verifies the presence of the inserted row in the underlying 'Sales.Customers' table, emphasizing the difference between the view's presentation and the actual data in the table.